### PR TITLE
planner/core: let mpp support partition tables

### PR DESCRIPTION
**Pull request by hanfei1991**
### What problem does this PR solve?

Problem Summary:
make mpp support partition tables.

### What is changed and how it works?

How it Works:
The principle is quite simple: alloc a task for every partition, so that every task can concentrate on a single "Physical Pairtition".

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below). Use tpc-h with paritition table.

### Release note 

- make mpp support paritition table.

### CODE REVIEW

File: executor/tiflash_test.go [hunk 1]
```diff
@@ -139,3 +139,56 @@ func (s *tiflashTestSuite) TestMppExecution(c *C) {
...
+	c.Assert(err, IsNil)
+
+	tk.MustExec("insert into t2 values(1,0)")
+	tk.MustExec("insert into t2 values(2,0)")
+	tk.MustExec("insert into t2 values(3,0)")
+	tk.MustExec("insert into t2 values(4,0)")
+	tk.MustQuery("select count(*) from t1 , t, t2 where t1.a = t.a and t2.a = t.a").Check(testkit.Rows("4"))
+
+	// test avg
+	tk.MustQuery("select avg(t1.a) from t1 , t where t1.a = t.a").Check(testkit.Rows("2.5000"))
```
Comment (fzhedu) [inline executor/tiflash_test.go:191]: should we ensure all query plans run in mpp?

File: executor/tiflash_test.go [hunk 2]
```diff
@@ -139,3 +139,56 @@ func (s *tiflashTestSuite) TestMppExecution(c *C) {
 	// test proj and selection
 	tk.MustQuery("select count(*) from (select a * 2 as a from t1) t1 , (select b + 4 as a from t)t where t1.a = t.a").Check(testkit.Rows("3"))
 }
+
+func (s *tiflashTestSuite) TestPartitionTable(c *C) {
```
Comment (fzhedu) [inline executor/tiflash_test.go:143]: 1. suggest to do a mannul test on tpch with different table partitions.
2. add some null data for integration tests
3. test some empty table for integration tests
4. test partitioned table works with non-partitioned tables.

File: planner/core/fragment.go [hunk 3]
```diff
@@ -110,4 +170,5 @@ func (e *mppTaskGenerator) constructMPPTasksImpl(ctx context.Context, ts *Physic
 		tasks = append(tasks, &kv.MPPTask{Meta: meta, ID: *e.allocTaskID, StartTs: e.startTS, TableID: tableID})
 	}
 	return tasks, nil
+
```
Comment (fzhedu) [inline planner/core/fragment.go:173]: delete this.

File: executor/tiflash_test.go (hunk 2 above)
Comment (hanfei1991) [reply executor/tiflash_test.go:143]: 1. The integration test for partition table which is based on TPC-H will be added after merging. For now, the manual test can work well.
2. good suggestion.
3. empty table is meaningless, but empty partition is nessesary, which have been included in the tests.
4. good suggestion.

File: executor/tiflash_test.go (hunk 1 above)
Comment (hanfei1991) [reply executor/tiflash_test.go:191]: Good question, I will test the plans as well

**Comment by fzhedu** [comment]
can we use the partition table to avoid exchange in some cases?

**Comment by hanfei1991** [comment]
@fzhedu Sure, but that's a future work

**Comment by fzhedu** [comment]
I care about the test for partition tables.
1. as described "Manual test (add detailed scripts or steps below). Use tpc-h with paritition table." does this run correctly?
2. can we port some partition tables from TiKV?

**Comment by hanfei1991** [comment]
1. yeah, run correctly.
2. Are there any integration tests for partition tables? I didn't find it.

File: planner/core/fragment.go [hunk 4]
```diff
@@ -87,19 +93,73 @@ func (e *mppTaskGenerator) generateMPPTasksForFragment(f *Fragment) (tasks []*kv
...
+	} else {
+		ret = make([]table.PhysicalTable, 0, len(idxArr))
+		for _, idx := range idxArr {
+			pid := pi.Definitions[idx].ID
+			p := tbl.GetPartition(pid)
+			ret = append(ret, p)
+		}
+	}
+	if len(ret) == 0 {
+		ret = []table.PhysicalTable{tbl.GetPartition(pi.Definitions[0].ID)}
```
Comment (qw4990) [inline planner/core/fragment.go:120]: Why use the first partition in this case that all partitions have been pruned?

**Comment by fzhedu** [comment]
> 2. Are there any integration tests for partition tables? I didn't find it.

1. https://github.com/pingcap/automated-tests/tree/master/cases/feature/table_partition
2. tidb/executor/partition_table

File: planner/core/fragment.go (hunk 4 above)
Comment (hanfei1991) [reply planner/core/fragment.go:120]: Because TiFlash cannot process a empty task correctly, then I choose to leave it with some data to read.

**Comment by hanfei1991** [comment]
@fzhedu for first test, we can migrate to TiFlash integration test.
for the unit test of partition table, it's to test the different tidb operator with parition tables. That's not suitable for tiflash test.

**Review by qw4990 — APPROVED** [review]

**Pull request merged**